catchpoint.py

```diff
@@ -27,53 +30,64 @@ def __init__(
         - host (str): The host to connect to
         - version (int): The version of the API
         """
+        self._client_id = client_id
+        self._client_secret = client_secret
         self._host = host
         self._version = version
-        self.content_type = "application/json"
 
         self._logger = logger if logger is not None else getLogger(name=self.__class__.__name__)
-        self._auth = False
-        self._token = None
+        self._headers = {
+            'Accept': 'application/json',
+        }
+        # Arbitrarily chosen date time in the past. Chosen so that a token will
+        # be requested before the first call.
+        self._token_expires_on = datetime(MINYEAR, 1, 1, tzinfo=pytz.utc)
 
     def _debug(self, msg):
         """
         Debug output. Set self.verbose to True to enable.
         """
         self._logger.debug(msg)
 
-    def _connection_error(self, e):
-        msg = "Unable to reach {0}: {1}".format(self._host, e)
-        sys.exit(msg)
-
-    def _authorize(self, creds):
+    def _get_header(self):
         """
         Request an auth token.
 
         - creds: dict with client_id and client_secret
         """
-        self._debug("Creating auth url...")
-        uri = "https://{0}/ui/api/token".format(self._host)
-        payload = {
-            'grant_type': 'client_credentials',
-            'client_id': creds['client_id'],
-            'client_secret': creds['client_secret']
-        }
+        now = datetime.now(tz=pytz.utc).replace(microsecond=0)
+        if self._token_expires_on < now:
+            # Remove old authentication header
+            if "Authorization" in self._headers:
+                del self._headers["Authorization"]
+
+            # Retrieve the new authentication token
+            self._debug("Creating auth url...")
+            uri = self._TOKEN_URL_TEMPLATE.format(host=self._host)
+            auth_token = self._make_request(
+                method="POST",
+                url=uri,
+                data={
+                    "grant_type": "client_credentials",
+                    "client_id": self._client_id,
+                    "client_secret": self._client_secret,
+                },
+                headers=self._headers,
+            )
 
-        # make request
-        self._debug("Making auth request...")
-        try:
-            r = requests.post(uri, data=payload)
-        except requests.ConnectionError as e:
-            self._connection_error(e)
+            # Set the new token in header
+            access_token = auth_token["access_token"]
+            self._debug("Access token: " + access_token)
+            self._headers["Authorization"] = "Bearer " + base64.b64encode(access_token)
 
-        self._debug("URL: " + r.url)
-        data = r.json()
+            # Remember the expiry datetime
+            # GOTCHA: Take out 60 seconds from the expiry just to be safe
```

Comment:
Can you explain this a little better than "just to be safe"? What is the gotcha you experienced here?